renderdoc/driver/gl/wrappers/gl_texture_funcs.cpp

```diff
@@ -1388,55 +1393,121 @@ void WrappedOpenGL::glCopyImageSubData(GLuint srcName, GLenum srcTarget, GLint s
                                  dstData.curType != eGL_TEXTURE_CUBE_MAP_ARRAY)
                                     ? RDCMAX(1, dstData.depth >> dstLevel)
                                     : dstData.depth;
-        if(srcX == 0 && srcY == 0 && srcZ == 0 && dstX == 0 && dstY == 0 && dstZ == 0 &&
-           srcLevelWidth == dstLevelWidth && srcLevelHeight == dstLevelHeight &&
-           srcLevelDepth == dstLevelDepth)
+
+        bytebuf srcCd;
+        // if we have source compressed data to copy
+        if(srcData.compressedData.find(srcLevel) != srcData.compressedData.end())
         {
-          dstData.compressedData[dstLevel] = srcData.compressedData[srcLevel];
+          // TODO: avoid copy
+          srcCd = srcData.compressedData[srcLevel];
         }
-        else
+        else if(!srcIsCompressed &&
+                (srcData.curType == eGL_TEXTURE_2D || srcData.curType == eGL_TEXTURE_2D_ARRAY))
         {
-          rdcfixedarray<uint32_t, 3> srcBlockSize = GetCompressedBlockSize(srcData.internalFormat);
-          rdcfixedarray<uint32_t, 3> dstBlockSize = GetCompressedBlockSize(dstData.internalFormat);
+          // try reading back without existing compressedData
+          RDCASSERT(!srcIsCompressed);
+
+          uint32_t size =
+              (uint32_t)GetByteSize(srcLevelWidth, srcLevelHeight, srcLevelDepth, srcFmt, srcType);
+          srcCd.resize(size);
 
-          size_t srcSliceSize = GetCompressedByteSize(srcLevelWidth, srcLevelHeight,
-                                                      srcBlockSize[2], srcData.internalFormat);
-          size_t dstSliceSize = GetCompressedByteSize(dstLevelWidth, dstLevelHeight,
-                                                      dstBlockSize[2], dstData.internalFormat);
+          GLuint prevTex = 0;
+          GL.glGetIntegerv(srcData.curType == eGL_TEXTURE_2D ? eGL_TEXTURE_BINDING_2D
+                                                             : eGL_TEXTURE_BINDING_2D_ARRAY,
+                           (GLint *)&prevTex);
+
+          GLenum oldActive = eGL_TEXTURE0;
+          GL.glGetIntegerv(eGL_ACTIVE_TEXTURE, (GLint *)&oldActive);
+          GL.glActiveTexture(eGL_TEXTURE0);
+
+          GL.glBindTexture(srcData.curType, srcName);
+          GL.glGetTexImage(srcData.curType, srcLevel, srcFmt, srcType, srcCd.data());
```

Comment:
I think you also need to push & pop the pixel pack state here too, to ensure tightly packed data that doesn't go to a buffer.